### DISCUSSION TITLE
几个初级的halo 2问题

### DISCUSSION THREAD

**Discussion by voidcenter** [z2o-k7e/halo2-step-by-step#4] [Q&A]
hello，我在看0xparc的fib[第一部分教程](https://learn.0xparc.org/materials/halo2/learning-group-1/halo2-api)，代码在[这里](https://github.com/icemelon/halo2-examples/blob/master/src/fibonacci/example1.rs)，有几个初级的问题哈

1. 这个AssignedCell<F,F>类型我没搞明白，为什么一个assigned cell会有两个类型，不应该只有一个类型吗？
2. 为啥这些给 region / advice / namespace 取名的地方都[用函数](https://github.com/icemelon/halo2-examples/blob/master/src/fibonacci/example1.rs#L103)而不用str，是什么情况真需要函数的灵活性还是这样在rust里面处理内存方便一点？ 
3. 我看[这里](https://github.com/icemelon/halo2-examples/blob/master/src/fibonacci/example1.rs#L39)的加法门从来没有被直接引用，是不是这个加法门就和[这个selector](https://github.com/icemelon/halo2-examples/blob/master/src/fibonacci/example1.rs#L31)绑死了，所以只要需要这个门就把这个selector enable，比如[这儿](https://github.com/icemelon/halo2-examples/blob/master/src/fibonacci/example1.rs#L68)。有多少种门就会有多少个selector，也就是说门之间不能共享selector？
4. constrain_instance 是不是底下也是 permutation argument，也就是说和 copy_advice 的原理是一样的，唯一的区别就是一个是public，一个不是public？

**Comment by Demian101** [comment]
1.
源码 Accept 2 个参数
```rust
/// An assigned cell.
#[derive(Clone, Debug)]
pub struct AssignedCell<V, F: Field> {
    value: Value<V>,
    cell: Cell,
    _marker: PhantomData<F>,
}
```

2. 你的意思可能是用 Closure ? 这边使用 namespace 的原因就是 debug 方便吧，It's purely for debugging. The namespaces are implemented via closures specifically so that in normal prover or verifier usage, they have zero cost (as the closures are never evaluated).

3. 这里是创建 custom gate，是在写”约束“，在 MockProver 内是可以看到被验证的，关于共享 selector 这个不知道，待解决.. 

4. 是的
比如 halo2_proofs/src/circuit/floor_planner/single_pass.rs  L203 :

```rust
    fn constrain_instance(
        &mut self,
        cell: Cell,
        instance: Column<Instance>,
        row: usize,
    ) -> Result<(), Error> {
        self.cs.copy(
            cell.column,
            *self.regions[*cell.region_index] + cell.row_offset,
            instance.into(),
            row,
        )
    }
```

**Reply by voidcenter** [reply]
1. 明白，就是说很多时候Cell value的类型可以不是 F 是吗？ 我以为 cell value的类型必须是 F，那什么情况下 cell value 类型不是 F 呢？
2. 👍
3. 如果我写了很多种不同的gate，在组建电路的时候该怎么选择用哪个gate呢？是不是每种gate都需要一列自己的selector，这样需要启动哪个gate就把哪列的selector enable？
4. 👍

**Reply by Demian101** [reply]
1. F 是个泛型，只要是满足 Trait 约束的任意类型都可以，这也是 generices 的强大之处

3. 在写 gate 的时候就需要标注好如何为这个 region 启用 Selector，这个只和你的约束逻辑有关； then , Layouter 会自动帮你做 regions 的布局

**Reply by voidcenter** [reply]
1. 我知道F是范型，只是我不知道 AssignedCell 为什么需要两个不同的范型 V 和 F.  V 这个好理解，F这个范型有什么用呢？   （其实PhantomData的作用我也一直不太清楚）

```
/// An assigned cell.
#[derive(Clone, Debug)]
pub struct AssignedCell<V, F: Field> {
    value: Value<V>,
    cell: Cell,
    _marker: PhantomData<F>,
}

```

关于2我是想问是不是每个gate必须独享一个select column，像这样：

<img width="735" alt="Screenshot 2023-10-27 at 1 02 32 AM" src="https://github.com/zkp-co-learning/halo2-step-by-step/assets/33541165/b624a19f-0a95-4e61-bc90-52d0b72d16c1">

还是说可以共享。似乎必须要独享，如果共享的话，把selector点亮就会把所有共享的gate启动，那等于这些gate合在一起是一个custom gate
